### DISCUSSION TITLE
Evaluate force field contributions of specific atom indexes

### DISCUSSION THREAD

**Discussion by NiklasKappel** [rdkit/rdkit#5579] [Ideas]
I constructed a force field using `RDKit::MMFF::constructForceField` for a molecule with, say, 10 atoms. For my application I would like to evaluate the energy contributed by a specific subset of the atoms of the molecule, say, the atoms with indexes `[0 1 8 9]`.

Browsing the source code, I see that force fields [sum over contributions](https://github.com/rdkit/rdkit/blob/master/Code/ForceField/ForceField.cpp#L323) and that contributions [know the atom indexes](https://github.com/rdkit/rdkit/blob/master/Code/ForceField/MMFF/TorsionAngle.h#L54) they care about. Also there seems to be the possibility to get the energy of every contribution [individually](https://github.com/rdkit/rdkit/blob/master/Code/ForceField/ForceField.cpp#L297).

For my molecule, I would like to sum over all contributions that care only about indexes in `[0 1 8 9]` and ignore contributions that care about any indexes in `[2 3 4 5 6 7]`. While force fields offer an interface to [get their contributions](https://www.rdkit.org/docs/cppapi/classForceFields_1_1ForceField.html#afbf46190b614f8906aae883d1259e026), contributions are [not required](https://www.rdkit.org/docs/cppapi/classForceFields_1_1ForceFieldContrib.html) to reveal their indexes.

The one way I can think of to make this possible would be to add a function `getIndexes` to the contribution Interface. However, I think that would mean that all 24 existing contribution types would have to be extended to fulfill that interface requirement.

Is there a more elegant way that I am not seeing? Ideally one that does net require changes to the rdkit code?

**Comment by ptosco** [comment]
@NiklasKappel I certainly would not claim it is an elegant way, but currently the only way to achieve what you describe without modifying the RDKit code would be to set verbosity to `MMFF::MMFF_VERBOSITY_HIGH` and redirect the output stream to a `std::stringstream`, then parse the latter to extract the contributions you are interested in.
The high verbosity logs should contain all the information you need.
Example below:
```C++
void testContribs() {
  auto molblock = R"(AMPTRB10
     RDKit          3D
AMPTRB10
 25 26  0  0  0  0  0  0  0  0999 V2000
   -8.1722   -2.5853   15.5036 N   0  3  0  0  0  0  0  0  0  0  0  0
   -7.3235   -2.9288   16.4553 C   0  0  0  0  0  0  0  0  0  0  0  0
   -6.9756   -2.0122   17.4036 N   0  0  0  0  0  0  0  0  0  0  0  0
   -6.7816   -4.1463   16.5340 N   0  0  0  0  0  0  0  0  0  0  0  0
   -7.1434   -5.0333   15.5874 C   0  0  0  0  0  0  0  0  0  0  0  0
   -6.5399   -6.2716   15.6657 N   0  0  0  0  0  0  0  0  0  0  0  0
   -8.0339   -4.7406   14.5489 C   0  0  0  0  0  0  0  0  0  0  0  0
   -8.3988   -5.6365   13.5936 N   0  0  0  0  0  0  0  0  0  0  0  0
   -9.2731   -5.2309   12.6368 C   0  0  0  0  0  0  0  0  0  0  0  0
   -9.6510   -6.2429   11.5962 C   0  0  0  0  0  0  0  0  0  0  0  0
   -9.7844   -3.9336   12.6374 C   0  0  0  0  0  0  0  0  0  0  0  0
  -10.7519   -3.4616   11.5914 C   0  0  0  0  0  0  0  0  0  0  0  0
   -9.4270   -3.0303   13.5932 N   0  0  0  0  0  0  0  0  0  0  0  0
   -8.5505   -3.4373   14.5480 C   0  0  0  0  0  0  0  0  0  0  0  0
   -8.5635   -1.6562   15.4710 H   0  0  0  0  0  0  0  0  0  0  0  0
   -7.2940   -1.0600   17.4772 H   0  0  0  0  0  0  0  0  0  0  0  0
   -6.3165   -2.3447   18.1027 H   0  0  0  0  0  0  0  0  0  0  0  0
   -5.9029   -6.4642   16.4265 H   0  0  0  0  0  0  0  0  0  0  0  0
   -6.8911   -7.0359   15.0983 H   0  0  0  0  0  0  0  0  0  0  0  0
  -10.7287   -6.4310   11.6198 H   0  0  0  0  0  0  0  0  0  0  0  0
   -9.1400   -7.1942   11.7780 H   0  0  0  0  0  0  0  0  0  0  0  0
   -9.3634   -5.8925   10.6003 H   0  0  0  0  0  0  0  0  0  0  0  0
  -11.0322   -2.4178   11.7673 H   0  0  0  0  0  0  0  0  0  0  0  0
  -11.6656   -4.0632   11.6167 H   0  0  0  0  0  0  0  0  0  0  0  0
  -10.3003   -3.5248   10.5968 H   0  0  0  0  0  0  0  0  0  0  0  0
  1 15  1  0  0  0  0
  1 14  2  0  0  0  0
  1  2  1  0  0  0  0
  2  4  2  0  0  0  0
  2  3  1  0  0  0  0
  3 17  1  0  0  0  0
  3 16  1  0  0  0  0
  4  5  1  0  0  0  0
  5  7  2  0  0  0  0
  5  6  1  0  0  0  0
  6 19  1  0  0  0  0
  6 18  1  0  0  0  0
  7 14  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  2  0  0  0  0
  9 11  1  0  0  0  0
  9 10  1  0  0  0  0
 10 22  1  0  0  0  0
 10 21  1  0  0  0  0
 10 20  1  0  0  0  0
 11 13  2  0  0  0  0
 11 12  1  0  0  0  0
 12 25  1  0  0  0  0
 12 24  1  0  0  0  0
 12 23  1  0  0  0  0
 13 14  1  0  0  0  0
M  CHG  1   1   1
M  END
)";
  std::unique_ptr<ROMol> mol(MolBlockToMol(molblock, true, false));
  TEST_ASSERT(mol);
  std::stringstream ss;
  MMFF::MMFFMolProperties mmffMolProperties(*mol, "MMFF94s", MMFF::MMFF_VERBOSITY_HIGH, ss);
  TEST_ASSERT(mmffMolProperties.isValid());
  std::unique_ptr<ForceFields::ForceField> ff(MMFF::constructForceField(*mol, &mmffMolProperties));
  TEST_ASSERT(ff);
  std::cout << ss.str() << std::endl;
}
```
```

**Comment by NiklasKappel** [comment]
For what it's worth I decided to implement my idea [here](https://github.com/rdkit/rdkit/compare/master...NiklasKappel:rdkit:partial_energy). If anyone cares I could make a pull request, but I had to make additions to every single force field contribution so I'm not sure if that is too invasive.
